### PR TITLE
Add parallel_reduce dynamic sized elements of View

### PR DESCRIPTION
**Pull request by Rombur**
With this PR, all the reduction tests are passing.

### CODE REVIEW

File: core/src/HIP/Kokkos_HIP_ReduceScan.hpp [hunk 1]
```diff
@@ -167,6 +167,257 @@ struct HIPReductionsFunctor<FunctorType, ArgTag, false, false> {
...
+  using ValueJoin   = FunctorValueJoin<FunctorType, ArgTag>;
+
+  using pointer_type = typename ValueTraits::pointer_type;
+
+  unsigned int const value_count   = ValueTraits::value_count(functor);
+  unsigned int const BlockSizeMask = hipBlockDim_y - 1;
+
+  // Must have power of two thread count
+
+  if (BlockSizeMask & blockDim.y) {
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_ReduceScan.hpp]: `hipBlockDim_y`?

File: core/src/HIP/Kokkos_HIP_ReduceScan.hpp [hunk 2]
```diff
@@ -167,6 +167,257 @@ struct HIPReductionsFunctor<FunctorType, ArgTag, false, false> {
...
+      };
+
+  auto block_scan_step = [&functor, value_count](pointer_type const TD,
+                                                 int const N, int const S) {
+    if (N == (1 << S)) {
+      ValueJoin::join(functor, TD, (TD - (value_count << S)));
+    }
+  };
+
+  const unsigned rtid_intra      = threadIdx.y ^ BlockSizeMask;
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_ReduceScan.hpp]: `hipThreadIdx_x`?

File: core/src/HIP/Kokkos_HIP_ReduceScan.hpp [hunk 3]
```diff
@@ -167,6 +167,257 @@ struct HIPReductionsFunctor<FunctorType, ArgTag, false, false> {
...
+  const pointer_type tdata_intra = base_data + value_count * threadIdx.y;
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_ReduceScan.hpp]: `hipThreadIdx_y`

File: core/src/HIP/Kokkos_HIP_ReduceScan.hpp [hunk 4]
```diff
@@ -167,6 +167,257 @@ struct HIPReductionsFunctor<FunctorType, ArgTag, false, false> {
...
+      }
+      if ((1 << 9) < BlockSizeMask) {
+        block_reduce_step(rtid_inter, tdata_inter, 9);
+      }
+      if ((1 << 10) < BlockSizeMask) {
+        block_reduce_step(rtid_inter, tdata_inter, 10);
+      }
+
+      int constexpr warp_size = ::Kokkos::Experimental::Impl::HIPTraits::WarpSize;
+      if (DoScan) {
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_ReduceScan.hpp:253]: Maybe comment that runtime check could be avoided.

File: core/src/HIP/Kokkos_HIP_ReduceScan.hpp [hunk 5]
```diff
@@ -167,6 +167,257 @@ struct HIPReductionsFunctor<FunctorType, ArgTag, false, false> {
...
+/**\brief  Input value-per-thread starting at 'shared_data'.
+ *         Reduction value at last thread's location.
+ *
+ *  If 'DoScan' then write blocks' scan values and block-groups' scan values.
+ *
+ *  Global reduce result is in the last threads' 'shared_data' location.
+ */
+
+template <bool DoScan, class FunctorType, class ArgTag>
+__device__ bool hip_single_inter_block_reduce_scan2(
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_ReduceScan.hpp:317]: I was going to complain about the name but I guess that is copied from Cuda :/

File: core/src/HIP/Kokkos_HIP_ReduceScan.hpp [hunk 6]
```diff
@@ -167,6 +167,257 @@ struct HIPReductionsFunctor<FunctorType, ArgTag, false, false> {
...
+  //        : (1 + atomicInc(global_flags, block_count - 1) < block_count));
+  __shared__ int n_done;
+  n_done = 0;
+  __syncthreads();
+  if (hipThreadIdx_y == 0) {
+    __threadfence();
+    n_done = 1 + atomicInc(global_flags, block_count - 1);
+  }
+  __syncthreads();
+  bool const is_last_block = n_done == block_count ? true : false;
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_ReduceScan.hpp]: Why the ternary operator?

**Review by dalg24 — CHANGES_REQUESTED** [review]

**Comment by Rombur** [comment]
Fixed

**Review by dalg24 — APPROVED** [review]

**Pull request merged**